File: examples/mathematics/jacobian_methods.rs
//! Compares different finite difference methods for numerical Jacobian computation.

#[allow(unused_imports)]
use brahe::math::jacobian::{DNumericalJacobian, DJacobianProvider};
use brahe::constants::GM_EARTH;
use brahe::R_EARTH;
use nalgebra::{DMatrix, DVector};

fn main() {
    // Define two-body gravity dynamics: state = [x, y, z, vx, vy, vz]
    let gravity_dynamics = |_t: f64, state: DVector<f64>| -> DVector<f64> {
        let r = state.fixed_rows::<3>(0);  // Position
        let v = state.fixed_rows::<3>(3);  // Velocity
        let r_norm = r.norm();

        // Acceleration from two-body gravity: a = -mu * r / |r|^3
        let a = -GM_EARTH * r / r_norm.powi(3);

        DVector::from_iterator(6, v.iter().chain(a.iter()).copied())
    };

    // Analytical Jacobian for two-body gravity
    let analytical_jacobian = |state: &DVector<f64>| -> DMatrix<f64> {
        let r = state.fixed_rows::<3>(0);
        let r_norm = r.norm();
        let r3 = r_norm.powi(3);
        let r5 = r_norm.powi(5);

        // Top-left: zeros (3x3)
        // Top-right: identity (3x3)
        // Bottom-left: gravity gradient (3x3)
        // Bottom-right: zeros (3x3)
        let mut jac = DMatrix::zeros(6, 6);

        // Velocity contribution to position derivative
        for i in 0..3 {
            jac[(i, 3 + i)] = 1.0;
        }

        // Gravity gradient term - Montenbruck Eqn 7.56
        let r_vec = r.into_owned();
        let gradient = -GM_EARTH * (DMatrix::identity(3, 3) / r3 - 3.0 * r_vec * r_vec.transpose() / r5);
        jac.view_mut((3, 0), (3, 3)).copy_from(&gradient);

        jac
    };

    // Create numerical Jacobians with different methods
    let jacobian_forward = DNumericalJacobian::forward(Box::new(gravity_dynamics));
    let jacobian_central = DNumericalJacobian::central(Box::new(gravity_dynamics));
    let jacobian_backward = DNumericalJacobian::backward(Box::new(gravity_dynamics));

    // Test state: Low Earth Orbit position and velocity
    let t = 0.0;
    let state = DVector::from_vec(vec![
        R_EARTH + 500e3,  // x position (m)
        0.0,               // y position
        0.0,               // z position
        0.0,               // x velocity
        7500.0,            // y velocity (m/s)
        0.0                // z velocity
    ]);

    // Compute analytical Jacobian
    let j_analytical = analytical_jacobian(&state);

    // Compute Jacobians with each method
    let j_forward = jacobian_forward.compute(t, state.clone());
    let j_central = jacobian_central.compute(t, state.clone());
    let j_backward = jacobian_backward.compute(t, state.clone());

    println!("Forward Difference Jacobian:");
    for i in 0..6 {
        print!("[");
        for j in 0..6 {
            print!("{:>10.2e}", j_forward[(i, j)]);
            if j < 5 { print!("  "); }
        }
        println!("]");
    }
    let error_forward = (j_forward.clone() - j_analytical.clone()).norm();
    println!("Error: {:.2e}\n", error_forward);

    println!("Central Difference Jacobian:");
    for i in 0..6 {
        print!("[");
        for j in 0..6 {
            print!("{:>10.2e}", j_central[(i, j)]);
            if j < 5 { print!("  "); }
        }
        println!("]");
    }
    let error_central = (j_central.clone() - j_analytical.clone()).norm();
    println!("Error: {:.2e}\n", error_central);

    println!("Backward Difference Jacobian:");
    for i in 0..6 {
        print!("[");
        for j in 0..6 {
            print!("{:>10.2e}", j_backward[(i, j)]);
            if j < 5 { print!("  "); }
        }
        println!("]");
    }
    let error_backward = (j_backward - j_analytical).norm();
    println!("Error: {:.2e}\n", error_backward);

    // Summary
    println!("Accuracy Comparison:");
    println!("  Forward:  {:.2e} (O(h))", error_forward);
    println!("  Central:  {:.2e} (O(h²))", error_central);
    println!("  Backward: {:.2e} (O(h))", error_backward);
    println!("\nCentral is {:.1}x more accurate than forward", error_forward / error_central);
    println!("Central is {:.1}x more accurate than backward", error_backward / error_central);
}

// Expected output:
// Forward Difference Jacobian:
// [    0.00e0      0.00e0      0.00e0      1.00e0      0.00e0      0.00e0]
// [    0.00e0      0.00e0      0.00e0      0.00e0      1.00e0      0.00e0]
// [    0.00e0      0.00e0      0.00e0      0.00e0      0.00e0      1.00e0]
// [   2.45e-6      0.00e0      0.00e0      0.00e0      0.00e0      0.00e0]
// [    0.00e0    -1.22e-6      0.00e0      0.00e0      0.00e0      0.00e0]
// [    0.00e0      0.00e0    -1.22e-6      0.00e0      0.00e0      0.00e0]
// Error: 5.54e-14

// Central Difference Jacobian:
// [    0.00e0      0.00e0      0.00e0      1.00e0      0.00e0      0.00e0]
// [    0.00e0      0.00e0      0.00e0      0.00e0      1.00e0      0.00e0]
// [    0.00e0      0.00e0      0.00e0      0.00e0      0.00e0      1.00e0]
// [   2.45e-6      0.00e0      0.00e0      0.00e0      0.00e0      0.00e0]
// [    0.00e0    -1.22e-6      0.00e0      0.00e0      0.00e0      0.00e0]
// [    0.00e0      0.00e0    -1.22e-6      0.00e0      0.00e0      0.00e0]
// Error: 3.40e-15

// Backward Difference Jacobian:
// [    0.00e0      0.00e0      0.00e0      1.00e0      0.00e0      0.00e0]
// [    0.00e0      0.00e0      0.00e0      0.00e0      1.00e0      0.00e0]
// [    0.00e0      0.00e0      0.00e0      0.00e0      0.00e0      1.00e0]
// [   2.45e-6      0.00e0      0.00e0      0.00e0      0.00e0      0.00e0]
// [    0.00e0    -1.22e-6      0.00e0      0.00e0      0.00e0      0.00e0]
// [    0.00e0      0.00e0    -1.22e-6      0.00e0      0.00e0      0.00e0]
// Error: 4.86e-14

// Accuracy Comparison:
//   Forward:  5.54e-14 (O(h))
//   Central:  3.40e-15 (O(h²))
//   Backward: 4.86e-14 (O(h))

// Central is 16.3x more accurate than forward
// Central is 14.3x more accurate than backward

File: examples/mathematics/jacobian_methods.py
# /// script
# dependencies = ["brahe", "numpy"]
# ///
"""
Compares different finite difference methods for numerical Jacobian computation.

This example demonstrates the accuracy trade-offs between forward, central,
and backward difference methods using two-body gravity dynamics.
"""

import brahe as bh
import numpy as np


# Define two-body gravity dynamics: state = [x, y, z, vx, vy, vz]
def gravity_dynamics(t, state):
    r = state[0:3]  # Position
    v = state[3:6]  # Velocity
    r_norm = np.linalg.norm(r)

    # Acceleration from two-body gravity: a = -mu * r / |r|^3
    a = -bh.GM_EARTH * r / r_norm**3

    return np.concatenate([v, a])


# Analytical Jacobian for two-body gravity
def analytical_jacobian(state):
    r = state[0:3]
    r_norm = np.linalg.norm(r)
    r3 = r_norm**3
    r5 = r_norm**5

    # Top-left: zeros (3x3)
    # Top-right: identity (3x3)
    # Bottom-left: gravity gradient (3x3)
    # Bottom-right: zeros (3x3)
    J = np.zeros((6, 6))
    J[0:3, 3:6] = np.eye(3)  # Velocity contribution to position derivative

    # Gravity gradient term - Motenbruck Eqn 7.56
    J[3:6, 0:3] = -bh.GM_EARTH * (np.eye(3) / r3 - 3 * np.outer(r, r) / r5)

    return J


# Create numerical Jacobians with different methods
jacobian_forward = bh.NumericalJacobian.forward(gravity_dynamics)
jacobian_central = bh.NumericalJacobian.central(gravity_dynamics)
jacobian_backward = bh.NumericalJacobian.backward(gravity_dynamics)

# Test state: Low Earth Orbit position and velocity
t = 0.0
state = np.array([bh.R_EARTH + 500e3, 0.0, 0.0, 0.0, 7500.0, 0.0])  # Circular orbit

# Compute analytical Jacobian
J_analytical = analytical_jacobian(state)

# Compute Jacobians with each method
J_forward = jacobian_forward.compute(t, state)
J_central = jacobian_central.compute(t, state)
J_backward = jacobian_backward.compute(t, state)

print("Forward Difference Jacobian:")
for row in J_forward:
    print("[" + "  ".join(f"{val: .2e}" for val in row) + "]")
error_forward = np.linalg.norm(J_forward - J_analytical)
print(f"Error: {error_forward:.2e}\n")

print("Central Difference Jacobian:")
for row in J_central:
    print("[" + "  ".join(f"{val: .2e}" for val in row) + "]")
error_central = np.linalg.norm(J_central - J_analytical)
print(f"Error: {error_central:.2e}\n")

print("Backward Difference Jacobian:")
for row in J_backward:
    print("[" + "  ".join(f"{val: .2e}" for val in row) + "]")
error_backward = np.linalg.norm(J_backward - J_analytical)
print(f"Error: {error_backward:.2e}\n")

# Summary
print("Accuracy Comparison:")
print(f"  Forward:  {error_forward:.2e} (O(h))")
print(f"  Central:  {error_central:.2e} (O(h²))")
print(f"  Backward: {error_backward:.2e} (O(h))")
print(f"\nCentral is {error_forward / error_central:.1f}x more accurate than forward")
print(f"Central is {error_backward / error_central:.1f}x more accurate than backward")

# Expected output:
# Forward Difference Jacobian:
# [ 0.00e+00   0.00e+00   0.00e+00   1.00e+00   0.00e+00   0.00e+00]
# [ 0.00e+00   0.00e+00   0.00e+00   0.00e+00   1.00e+00   0.00e+00]
# [ 0.00e+00   0.00e+00   0.00e+00   0.00e+00   0.00e+00   1.00e+00]
# [ 2.45e-06   0.00e+00   0.00e+00   0.00e+00   0.00e+00   0.00e+00]
# [ 0.00e+00  -1.22e-06   0.00e+00   0.00e+00   0.00e+00   0.00e+00]
# [ 0.00e+00   0.00e+00  -1.22e-06   0.00e+00   0.00e+00   0.00e+00]
# Error: 5.54e-14

# Central Difference Jacobian:
# [ 0.00e+00   0.00e+00   0.00e+00   1.00e+00   0.00e+00   0.00e+00]
# [ 0.00e+00   0.00e+00   0.00e+00   0.00e+00   1.00e+00   0.00e+00]
# [ 0.00e+00   0.00e+00   0.00e+00   0.00e+00   0.00e+00   1.00e+00]
# [ 2.45e-06   0.00e+00   0.00e+00   0.00e+00   0.00e+00   0.00e+00]
# [ 0.00e+00  -1.22e-06   0.00e+00   0.00e+00   0.00e+00   0.00e+00]
# [ 0.00e+00   0.00e+00  -1.22e-06   0.00e+00   0.00e+00   0.00e+00]
# Error: 5.27e-15

# Backward Difference Jacobian:
# [ 0.00e+00   0.00e+00   0.00e+00   1.00e+00   0.00e+00   0.00e+00]
# [ 0.00e+00   0.00e+00   0.00e+00   0.00e+00   1.00e+00   0.00e+00]
# [ 0.00e+00   0.00e+00   0.00e+00   0.00e+00   0.00e+00   1.00e+00]
# [ 2.45e-06   0.00e+00   0.00e+00   0.00e+00   0.00e+00   0.00e+00]
# [ 0.00e+00  -1.22e-06   0.00e+00   0.00e+00   0.00e+00   0.00e+00]
# [ 0.00e+00   0.00e+00  -1.22e-06   0.00e+00   0.00e+00   0.00e+00]
# Error: 6.59e-14

# Accuracy Comparison:
#   Forward:  5.54e-14 (O(h))
#   Central:  5.27e-15 (O(h²))
#   Backward: 6.59e-14 (O(h))

# Central is 10.5x more accurate than forward
# Central is 12.5x more accurate than backward
